fili-core/src/main/java/com/yahoo/bard/webservice/web/handlers/CacheWeightCheckRequestHandler.java

```diff
@@ -0,0 +1,225 @@
+
+    /**
+     * Build a cache weight checking request handler.
+     *
+     * @param next  The request handler to delegate the request to.
+     * @param webService  The web service to use for weight checking
+     * @param queryWeightUtil  A provider which measures estimated weight against allowed weights.
+     * @param mapper  A JSON object mapper, used to parse the JSON response from the weight check.
+     * @param dataCache The cache instance
+     * @param querySigningService The service to generate query signatures
+     */
+    public CacheWeightCheckRequestHandler(
+            DataRequestHandler next,
+            DruidWebService webService,
+            QueryWeightUtil queryWeightUtil,
+            ObjectMapper mapper,
+            @NotNull DataCache<?> dataCache,
+            QuerySigningService<?> querySigningService
+    ) {
+        super(
+                next,
+                webService,
+                queryWeightUtil,
+                mapper
+        );
+        this.dataCache = (TupleDataCache<String, Long, String>) dataCache;
+        this.querySigningService = (QuerySigningService<Long>) querySigningService;
+    }
+
+    @Override
+    public boolean handleRequest(
+            final RequestContext context,
+            final DataApiRequest request,
+            final DruidAggregationQuery<?> druidQuery,
+            final ResponseProcessor response
+    ) {
+        // Heuristic test to let requests with very low estimated cardinality directly through
+        if (queryWeightUtil.skipWeightCheckQuery(druidQuery)) {
+            return next.handleRequest(context, request, druidQuery, response);
+        }
+
+        BardQueryInfo.getBardQueryInfo().incrementCountWeightCheck();
+        final WeightCheckResponseProcessor weightCheckResponse =
+                new WeightCheckResponseProcessor(response);
+        final DruidAggregationQuery<?> weightEvaluationQuery = queryWeightUtil.makeWeightEvaluationQuery(druidQuery);
+        Granularity granularity = druidQuery.getInnermostQuery().getGranularity();
+        final long queryRowLimit = queryWeightUtil.getQueryWeightThreshold(granularity);
+
+        try {
+            LOG.debug("Weight query {}", writer.writeValueAsString(weightEvaluationQuery));
+        } catch (JsonProcessingException e) {
+            LOG.warn("Weight Query json exception:", e);
+        }
+
+        CacheService cacheService = new CacheService();
+        String cacheKey = null;
+        try {
+            cacheKey = getKey(druidQuery, mapper, writer);
+        } catch (JsonProcessingException e) {
+            LOG.warn("Cache key cannot be built: ", e);
+        }
+
+        WeightCheckRequestHandler weightCheckRequestHandler =
+                new WeightCheckRequestHandler(next, webService, queryWeightUtil, mapper);
+        SuccessCallback weightQuerySuccess = null;
+        // Construct success callback based on cache hit or cache miss
+        if (context.isReadCache()) {
```

Comment:
We can flatten the nesting if we handle the not case of this first:
```suggestion
        // No possible cache hit, so just run the query and save to cache
        if (!context.isReadCache()) {
            webService.postDruidQuery(context, cahingSuccessCallback, error, failure, weightEvaluationQuery);
            return true;
        }
```